Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/trace.py

```diff
@@ -0,0 +1,276 @@
+        'cherry-picked', 'original',
+        'reverts', 'reverted by',
+        'follow-up', 'follow-up by',
+    )
+    REFERENCES, REFERENCED_BY, \
+    CHERRY_PICK, ORIGINAL, \
+    REVERTS, REVERTED_BY, \
+    FOLLOW_UP, FOLLOW_UP_BY = TYPES
+
+    PAIRED = [FOLLOW_UP, FOLLOW_UP_BY]
+    IDENTITY = [CHERRY_PICK, ORIGINAL]
+    UNDO = [REVERTS, REVERTED_BY]
+
+    @classmethod
+    def reversed(cls, type):
+        return {
+            cls.REFERENCES: cls.REFERENCED_BY,
+            cls.REFERENCED_BY: cls.REFERENCES,
+            cls.CHERRY_PICK: cls.ORIGINAL,
+            cls.ORIGINAL: cls.CHERRY_PICK,
+            cls.REVERTS: cls.REVERTED_BY,
+            cls.REVERTED_BY: cls.REVERTS,
+            cls.FOLLOW_UP: cls.FOLLOW_UP_BY,
+            cls.FOLLOW_UP_BY: cls.FOLLOW_UP,
+        }.get(type, type)
+
+    @classmethod
+    def parse(cls, commit):
+        lines = commit.message.splitlines()
+
+        for type, regexes in {
+            cls.ORIGINAL: [CHERRY_PICK_RE],
+            cls.REVERTS: [REVERT_RE],
+            cls.FOLLOW_UP: FOLLOW_UP_FIXES_RE,
+        }.items():
+            for regex in regexes:
+                match = regex.match(lines[0])
+                if not match:
+                    continue
+                primary = match.group('primary')
+                secondary = match.group('secondary')
+                if secondary:
+                    secondary = UNPACK_SECONDARY_RE.match(secondary).groups()[0]
+                return type, [ref.rstrip('.').rstrip() for ref in [primary, secondary] if ref]
+        return None, []
+
+    def __init__(self, commit, type=None):
+        self.commit = commit
+        self.type = type or self.REFERENCES
+
+    def __repr__(self):
+        return '{} {}'.format(self.commit, self.type)
+
+
+class CommitsStory(object):
+    @classmethod
+    def issues_for(cls, commit):
+        filter = set()
+        result = []
+        line_count = 0
+        for line in commit.message.splitlines():
+            if not line and line_count >= 2:
+                break
+            if not line:
+                continue
+            line_count += 1
+            for word in line.split(' '):
+                bug = Tracker.from_string(word)
+                if bug and bug.link not in filter:
+                    filter.add(bug.link)
+                    result.append(bug)
+        return result
+
+    def __init__(self, commits=None):
+        self.commits = []
+        self.by_ref = {}
+        self.by_issue = {}
+        self.relations = {}
+        for commit in commits or []:
+            self.add(commit)
+
+    def add(self, commit):
+        if str(commit) in self.by_ref:
+            return True
+        self.commits.append(commit)
+        self.by_ref[str(commit)] = commit
+        if commit.hash:
+            self.by_ref[commit.hash[:Commit.HASH_LABEL_SIZE]] = commit
+        if commit.revision:
+            self.by_ref['r{}'.format(commit.revision)] = commit
```

Comment:
We're still writing tooling to deal with Trac-style revisions?

Reply:
Not aggressively, but it's conceivable this tool (`git-webkit trace`, not `git-webkit pickable`) be use to deal with commits pre-transition.